### PR TITLE
Element market support

### PR DESCRIPTION
**Pull request by hensha256**

### CODE REVIEW

File: test/forge/SwapNFTCallParameters.t.sol [hunk 1]
```diff
@@ -63,6 +63,28 @@ contract swapNFTCallParametersTest is Test, Interop, DeployRouter {
         assertEq(from.balance, balance - params.value);
     }
 
+    function testElementBuyItemsERC721() public {
+        address taker = 0x75B6568025f463a98fB01082eEb6dCe04efA3Ae4;
```
Comment (hensha256) [inline test/forge/SwapNFTCallParameters.t.sol:67]: having trouble with their API right now to use our `TEST_RECIPIENT` as the taker. Will add an issue to the repo to update this for consistency :)

File: src/entities/protocols/element-market.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        item.order,
+        item.signature,
+        item.order.taker, // taker
+        '0x', // extraData
+      ])
+
+      planner.addCommand(CommandType.ELEMENT_MARKET, [value.toString(), calldata], config.allowRevert)
+      if(item.order.taker == ElementTrade.SENDER_AS_TAKER) {
+        // TODO: we need to add an input for the eventual recipient of the NFT
+        planner.addCommand(CommandType.SWEEP_ERC721, [item.order.nft, item.order.taker, item.order.nftId])
```
Comment (zhongeric) [inline src/entities/protocols/element-market.ts:83]: @hensha256 We need to allow for the SDK to also bundle `SWEEP_ERC721` after the call to Element as for most orders fetched through their API, the `taker` field is the zero address which just sends the NFT to sender (or UR). I'm not sure what the best way of doing that is / if we have precedent for that.

File: src/entities/protocols/element-market.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  applyFeesToAmount(erc20TokenAmount: string, fees: Fee[]): string {
+    if(fees) {
+      const feeSum = fees.reduce((acc, fee) => {
+        return acc.add(BigNumber.from(fee.amount))
+      }, BigNumber.from(0))
+      if(feeSum.gt(erc20TokenAmount)) throw new Error('Fees cannot be greater than order amount')
+      erc20TokenAmount = BigNumber.from(erc20TokenAmount).add(feeSum).toString()
+    }
+    return erc20TokenAmount
+  }
```
Comment (zhongeric) [inline src/entities/protocols/element-market.ts:64]: This is new, and it's super annoying that they wrote their contracts this way but basically in the calldata we have to send the original erc20amount that the user signed off on, BUT for the value that we send, we have to include more for fees.

**Review by zhongeric — CHANGES_REQUESTED** [review]

File: src/entities/protocols/element-market.ts (hunk 2 above)
Comment (hensha256) [reply src/entities/protocols/element-market.ts:83]: And we can't fill in the taker field with the correct recipient?

**Review by zhongeric — APPROVED** [review]

**Pull request merged**